rag-toolchain/src/retrievers/postgres_vector_retriever.rs

```diff
@@ -0,0 +1,77 @@
+use crate::clients::traits::AsyncEmbeddingClient;
+use crate::common::types::Chunk;
+use crate::retrievers::traits::AsyncRetriever;
+use async_trait::async_trait;
+use pgvector::Vector;
+use sqlx::postgres::PgRow;
+use sqlx::Error as SqlxError;
+use sqlx::{Pool, Postgres, Row};
+use std::error::Error;
+use std::fmt::{Display, Formatter};
+
+pub struct PostgresVectorRetriever<T>
+where
+    T: AsyncEmbeddingClient,
+{
+    pub pool: Pool<Postgres>,
+    table_name: String,
+    embedding_client: T,
+}
+
+impl<T: AsyncEmbeddingClient> PostgresVectorRetriever<T> {
+    pub fn new(pool: Pool<Postgres>, table_name: String, embedding_client: T) -> Self {
+        PostgresVectorRetriever {
+            pool,
+            table_name,
+            embedding_client,
+        }
+    }
+}
+
+#[async_trait]
+impl<T> AsyncRetriever for PostgresVectorRetriever<T>
+where
+    T: AsyncEmbeddingClient + Sync,
+    T::ErrorType: 'static,
+{
+    type ErrorType = PostgresRetrieverError<T::ErrorType>;
+    async fn retrieve(&self, text: &str) -> Result<Chunk, Self::ErrorType> {
+        let (_, embedding) = self
+            .embedding_client
+            .generate_embedding(text.into())
+            .await
+            .map_err(PostgresRetrieverError::EmbeddingClientError)?;
+        let mapped_embedding: Vector = Vector::from(embedding.embedding().to_vec());
+        let embedding_query: String = format!(
+            "
+            SELECT content FROM {} ORDER BY embedding <=> $1 LIMIT 1",
```

Comment:
I think we can add the cast back ie ::vector instead of having to use the vector type ?